File: mobile/src/main/scala/com/github/shadowsocks/preference/IconListPreference.scala
package com.github.shadowsocks.preference

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.annotation.ArrayRes
import android.support.v7.preference.ListPreference
import android.support.v7.preference.Preference.OnPreferenceChangeListener
import android.util.AttributeSet

/**
  * @author Mygod
  */
class IconListPreference(context: Context, attrs: AttributeSet = null) extends ListPreference(context, attrs) {
  private var mEntryIcons: Array[Drawable] = _
  def selectedEntry: Int = getEntryValues.indexOf(getValue)

  var unknownValueSummary: String = _
  var entryPackageNames: Array[String] = _

  private var listener: OnPreferenceChangeListener = _
  override def getOnPreferenceChangeListener: OnPreferenceChangeListener = listener
  override def setOnPreferenceChangeListener(listener: OnPreferenceChangeListener): Unit = this.listener = listener
  super.setOnPreferenceChangeListener((preference, newValue) =>
    if (listener == null || listener.onPreferenceChange(preference, newValue)) {
      setValue(newValue.toString)
      checkSummary()
      if (mEntryIcons != null) setIcon(getEntryIcon)
      true
    } else false)
  def checkSummary(): Unit =
    if (unknownValueSummary != null) setSummary(if (selectedEntry < 0) unknownValueSummary.format(getValue) else "%s")

//  {
//    val a: TypedArray = context.obtainStyledAttributes(attrs, R.styleable.IconListPreference)
//    val entryIconsResId: Int = a.getResourceId(R.styleable.IconListPreference_entryIcons, -1)
//    if (entryIconsResId != -1) setEntryIcons(entryIconsResId)
//    a.recycle()
//  }

  def getEntryIcon: Drawable = try mEntryIcons(selectedEntry) catch {
    case _: ArrayIndexOutOfBoundsException => null
  }
  def getEntryIcons: Array[Drawable] = mEntryIcons
  def setEntryIcons(entryIcons: Array[Drawable]): Unit = mEntryIcons = entryIcons

  def setEntryIcons(@ArrayRes entryIconsResId: Int) {
    val icons_array = getContext.getResources.obtainTypedArray(entryIconsResId)
    val icon_ids_array = new Array[Drawable](icons_array.length)
    for (i <- 0 until icons_array.length) icon_ids_array(i) = icons_array.getDrawable(i)
    setEntryIcons(icon_ids_array)
    icons_array.recycle()
  }

  def init() {
    val entryValues = getEntryValues
    if (entryValues == null) return
    if (mEntryIcons != null) setIcon(getEntryIcon)
  }

  protected override def onSetInitialValue(restoreValue: Boolean, defaultValue: AnyRef) {
    super.onSetInitialValue(restoreValue, defaultValue)
    init()
  }
}

File: mobile/src/main/java/com/github/shadowsocks/preference/IconListPreference.kt
/*******************************************************************************
 *                                                                             *
 *  Copyright (C) 2017 by Max Lv <max.c.lv@gmail.com>                          *
 *  Copyright (C) 2017 by Mygod Studio <contact-shadowsocks-android@mygod.be>  *
 *                                                                             *
 *  This program is free software: you can redistribute it and/or modify       *
 *  it under the terms of the GNU General Public License as published by       *
 *  the Free Software Foundation, either version 3 of the License, or          *
 *  (at your option) any later version.                                        *
 *                                                                             *
 *  This program is distributed in the hope that it will be useful,            *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 *  GNU General Public License for more details.                               *
 *                                                                             *
 *  You should have received a copy of the GNU General Public License          *
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *                                                                             *
 *******************************************************************************/

package com.github.shadowsocks.preference

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v7.preference.ListPreference
import android.util.AttributeSet

class IconListPreference(context: Context, attrs: AttributeSet? = null) : ListPreference(context, attrs) {
    var entryIcons: Array<Drawable?>? = null
    val selectedEntry: Int get() = entryValues.indexOf(value)
    val entryIcon: Drawable? get() = try {
        entryIcons?.get(selectedEntry)
    } catch (_: ArrayIndexOutOfBoundsException) {
        null
    }
//    fun setEntryIcons(@ArrayRes entryIconsResId: Int) {
//        val array = getContext().getResources().obtainTypedArray(entryIconsResId)
//        entryIcons = Array(array.length(), { i -> array.getDrawable(i) })
//        array.recycle()
//    }

    var unknownValueSummary: String? = null
    var entryPackageNames: Array<String>? = null

    private var listener: OnPreferenceChangeListener? = null
    override fun getOnPreferenceChangeListener(): OnPreferenceChangeListener? = listener
    override fun setOnPreferenceChangeListener(listener: OnPreferenceChangeListener?) {
        this.listener = listener
    }

    init {
        super.setOnPreferenceChangeListener({ preference, newValue ->
            val listener = listener
            if (listener == null || listener.onPreferenceChange(preference, newValue)) {
                value = newValue.toString()
                checkSummary()
                if (entryIcons != null) icon = entryIcon
                true
            } else false
        })
//        val a = context.obtainStyledAttributes(attrs, R.styleable.IconListPreference)
//        val entryIconsResId: Int = a.getResourceId(R.styleable.IconListPreference_entryIcons, -1)
//        if (entryIconsResId != -1) entryIcons = entryIconsResId
//        a.recycle()
    }

    fun checkSummary() {
        val unknownValueSummary = unknownValueSummary
        if (unknownValueSummary != null) summary = if (selectedEntry < 0) unknownValueSummary.format(value) else "%s"
    }

    fun init() {
        icon = entryIcon
    }
    override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any?) {
        super.onSetInitialValue(restoreValue, defaultValue)
        init()
    }
}
